NuRadioReco/modules/io/coreas/coreas.py

```diff
@@ -1,140 +1,328 @@
+    # second to last dimension has to be 3 for the transformation
+    obs_positions_geo = cs.transform_from_magnetic_to_geographic(obs_positions.T)
 
-        if debug:
-            ax2.plot(vertices_ground[:, 0], vertices_ground[:, 1], c='grey', zorder=1)
-            ax2.scatter(vertices_ground[:, 0], vertices_ground[:, 1], c='tab:orange', zorder=2)
-    if debug:
-        ax2.scatter(positions[:, 0], positions[:, 1], c='tab:blue', s=10, label='Position of observer')
-        ax2.scatter(vertices_ground[:, 0], vertices_ground[:, 1], c='tab:orange', label='Vertices of cell')
-        ax2.set_aspect('equal')
-        ax2.set_title('On ground, total area {:.2f} $km^2$'.format(sum(weights) / units.km ** 2))
-        ax2.set_xlabel('East [m]')
-        ax2.set_ylabel('West [m]')
-        ax2.set_xlim(-5000, 5000)
-        ax2.set_ylim(-5000, 5000)
-        plt.legend()
-        plt.show()
+    return obs_positions_geo
 
-        fig3 = plt.figure(figsize=[12, 4])
-        ax4 = fig3.add_subplot(121)
-        ax5 = fig3.add_subplot(122)
 
-        ax4.hist(weights)
-        ax4.set_title('Weight distribution')
-        ax4.set_xlabel(r'Weights (here area) $[m^2]$')
-        ax4.set_ylabel(r'Number of observer')
+def convert_obs_positions_to_vxB_vxvxB(observer, zenith, azimuth, magnetic_field_vector):
+    """
+    Converts one observer positions from the corsika file to NuRadio units in the (vxB, vxvxB) shower plane.
+    coreas: x-axis pointing to the magnetic north, the positive y-axis to the west, and the z-axis upwards.
+    NuRadio: x-axis pointing to the east, the positive y-axis geographical north, and the z-axis upwards.
+    NuRadio_x = -coreas_y, NuRadio_y = coreas_x, NuRadio_z = coreas_z and then correct for mag north
 
-        ax5.scatter(length_shower ** 2, weights)
-        ax5.set_xlabel(r'$Length^2 [m^2]$')
-        ax5.set_ylabel('Weight $[m^2]$')
-        plt.show()
+    Parameters
+    ----------
+    observer : value
+        observer as in the hdf5 file object, e.g. corsika['CoREAS']['observers'].values()[i]
+    zenith : float
+        zenith angle in radians from corsika file, e.g. use get_angles()
+    azimuth : float
+        azimuth angle in radians from corsika file, e.g. use get_angles()
+    magnetic_field_vector : np.array (3)
+        magnetic field vector from corsika file, e.g. use get_angles()
+  
+    Returns
+    -------
+    obs_positions_vxB_vxvxB: np.array (3)
+        observer positions in (vxB, vxvxB) shower plane
+    
+    """
+    cs = coordinatesystems.cstrafo(zenith, azimuth, magnetic_field_vector)
 
-    return weights
+    obs_positions_geo = convert_obs_positions_to_nuradio_on_ground(observer, zenith, azimuth, magnetic_field_vector)
+
+    # transforms the coreas observer positions into the vxB, vxvxB shower plane
+    obs_positions_vxB_vxvxB = cs.transform_to_vxB_vxvxB(obs_positions_geo).T
+
+    return obs_positions_vxB_vxvxB
+
+
+def read_CORSIKA7(input_file, declination=None):
```

Comment:
Can someone help me understand why the declination parameter is necessary here? Does it not come from the input_file itself?

Reply:
Without having checked the code and what the function does with the declination: No the declination does not come from the input_file. The normal corsika does not care about the declination at all as the coordinate system is defined by the magnetic north. There is the possibility to store the declination as meta data in an `*reas` file but it wont effect the simulation at all (its only intended to be stored along a simulation to use it once the simulation is converted to a different coordinate system in an reconstruction framework). Did that help?

Reply:
I see. But one could get it from the radiotools helper function, which has the 3D magnetic field, right? Do I understand that correctly?